Add SMA EV Charger

powelllens opened the pull request.

powelllens committed:
New Version of MaxCurrent for charger
- add support for 10mA steps MaxCurrentMillis
- changed parameters / behavior of current limiter

powelllens committed:
-add auth  file /smaevcharger/identity.go
-changed Token Auth functionality Token will be used until expired, refresh of token via access token is currently not working. As a Workaround we acquire a new token with the username and password.

-moved const to /smaevcharger/types.go

-cleaned up several functions and used already existing helpers

-add multi parameter upload functionalities

-some minor changes and bugfixes

powelllens committed:
- changes Software to support Charger Software version 1.2.23R
- due to changed behavior earlier Software will not be supported
- add software version control during startup
- Current control workaround no longer needed, current is now be controlled with the AC Line Current Limiter

powelllens committed:
rebase to master

powelllens committed:
upload files to new branch

andig committed:
wip

powelllens commented on charger/smaevcharger/identity.go:
```diff
@@ -0,0 +1,116 @@
+func (c *tokenSource) RefreshToken(token *oauth2.Token) (*oauth2.Token, error) {
+	/*
+
+		Uri := c.Host + "/refresh_token" //URL Unknown?!
```
refresh URL unknown, no api documentation for token refresh avaliable

powelllens commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+	if enable {
+		switch StateChargerSwitch {
+		case smaevcharger.ConstSwitchOeko: // Switch PV Loading
+			wb.SendParameter("Parameter.Chrg.ActChaMod", smaevcharger.ConstOptiCharge)
```
Currently no error message is returned if the charger is not set to fast charge.

powelllens replied:
error message was added

premultiply commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+		return fmt.Errorf("invalid current %v", current)
+	}
+
+	wb.SendParameter("Parameter.Inverter.AcALim", strconv.FormatInt(int64(current), 10))
+	time.Sleep(time.Second)
+	return nil
```
```suggestion
	return wb.MaxCurrentMillis(float64(current))
```

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+		return wb, errors.New(errortext)
+	}
+
+	if tempvar1 < 1 || tempvar2 < 2 || tempvar3 < 23 {
```
Das ist sehr fragil. `2.1.0` failt. `semver` Package benutzen!

powelllens replied:
fixed, used recommended lib

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+	}
+
+	wb.SendParameter("Parameter.Inverter.AcALim", strconv.FormatInt(int64(current), 10))
+	time.Sleep(time.Second)
```
Warum?

andig replied:
strconv.FormatInt(int64(current), 10)

ist doch

    strconv.Itoa

?

powelllens replied:
obsolet da durch andere Funktion ersetzt

andig replied:
Du musst nichts kommentieren, GH macht die Kommentare automatisch zu wenn der Code sich ändert.

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+		return fmt.Errorf("invalid current %.5g", current)
+	}
+	wb.SendParameter("Parameter.Inverter.AcALim", strconv.FormatFloat(current, 'f', 2, 64))
+	time.Sleep(time.Second)
```
Dito

powelllens replied:
für float64 zu string habe ich keine alternative gefunden, Itoa gilt nur für Int?

andig replied:
fmt.Sprintf

gäbs noch, aber hast recht

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+func (wb *Smaevcharger) GetMeasurement(id string) interface{} {
+	dataok := wb.GetMeasurementData()
+	if !dataok {
+		nil := fmt.Errorf("failed to aquire measurement data")
```
Ui- das compiliert?

powelllens replied:
return fixed / changed

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+	var payload smaevcharger.SendData
+
+	for i := range data {
+		payload.Timestamp = time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
```
Format als Konstante

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+	return PhsA, PhsB, PhsC, nil
+}
+
+func (wb *Smaevcharger) GetMeasurementData() bool {
```
Sollten wir hier nicht mit `updated`/`timeout` arbeiten?

andig commented on templates/definition/charger/smaevcharger.yaml:
```diff
@@ -0,0 +1,42 @@
+      en: Will be displayed in the user interface
+  - name: host
+    required: true
+    description:
```
Die Beschreibungen bitte raus, die sind schon per Default definiert

andig commented on templates/definition/charger/smaevcharger.yaml:
```diff
@@ -0,0 +1,42 @@
+      en: IP-Address or Hostname
+    example: 192.0.2.2
+  - name: user
+    description:
```
dito

andig commented on templates/definition/charger/smaevcharger.yaml:
```diff
@@ -0,0 +1,42 @@
+      de: Benutzerkonto (bspw. E-Mail Adresse, User Id, etc.)
+      en: User-account (e.g. email address, user id, etc.)
+  - name: password
+    description:
```
dito

andig commented on templates/definition/charger/smaevcharger.yaml:
```diff
@@ -0,0 +1,42 @@
+      generic: Wallbox
+params:
+  - name: title
+    description:
```
M.E. fragen wir title bei den Chargern nicht ab?

powelllens committed:
Update charger/smaevcharger.go

Co-authored-by: premultiply <ulrich.peters@posteo.de>

powelllens committed:
-moved time format to const
-removed control if modified parameter is avaliable, sending not existing parameter will fail anyway
-removed no longer needed comments in identity.go

powelllens committed:
Merge branch 'addWallboxSMAEVCharger' of https://github.com/powelllens/evcc into addWallboxSMAEVCharger

powelllens committed:
-removed input current control / will be checked in called function MaxCurrentMillis
-fixed bad return message
-cleaned up if / else
-removed all comments and descriptions from templates/definition/charger/smaevcharger.yaml

powelllens commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,359 @@
+		nil := fmt.Errorf("failed to aquire measurement data")
+		return nil
+	}
+	var returndata interface{}
```
erledigt

powelllens committed:
-changed software version control during startup to something more robust

powelllens committed:
-add version control errorhandling
-add error message if switch position is not in fast charge mode during activation

powelllens committed:
-convert error messages to lower case
-removed optional from template

powelllens committed:
- moved min accepted software version as const to types.go

powelllens committed:
- add cache and duration handling for data acquisition
- improved / changed error handling for most functions - fixed broken error handling

powelllens committed:
-fixed formatting
-minor bugfix

powelllens committed:
-add missing states to types.go (locked charger/locking via app/manual override)
-finally fixed token refresh and removed workaround
-some more bugfixes to charger behavior (this should mainly fix all "out of sync errors"):
--override status during startup and set needed parameters to prevent unreadable status (thanks SMA)
--add oldstate to compare old with new status to override charge mode if car gets connected, to prevent auto charge start and let EVCC decide what to do (Charger seems to change the charge state by many different events automatically)

andig committed:
Simplify

andig commented:
@powelllens ich habe einmal komplett vereinfacht und eine Reihe von Todos ergänzt. Das Handling von dem getPara/Measurements ist noch ziemlich unglücklich. Schau mal bitte wie das sonst gemacht ist- besser wäre jeweils Wert und Error zurück zu geben. Ganz fancy könntest Du sogar ein `provider.Cached` benutzen.

andig committed:
Simplify caching

andig commented:
Update: ich hab das Cachig auch noch vereinfach- magst Du noch die letztes TODOs anschauen?

andig committed:
simplify

andig committed:
Add error handling

andig committed:
Template

andig commented:
Hab auch das noch geändert. Bitte schau nochmal rein. Hoffe es tut noch...

andig committed:
wip

andig committed:
wip

andig committed:
Simplify user/password

andig committed:
wip

powelllens committed:
fixed TODO's
fixed uri bug after redesign / simplification
removed charger status override during startup
fixed out of sync bug to not override Charging Mode if new Status is StatusB but old Status was not StatusA - forced update after override
fixed bug in Send() "EOF" - DoJSon fails -> Http response without any Body
returned to status check

powelllens commented:
Habe alle Fehler behoben die durch das redesign / simplifycation passiert sind, läuft jetzt wieder.
Die TODO's habe ich soweit ergänzt / geändert.
+ Ein paar Fehlerbehebungen in der Statusmeldung
+ Alle Wartezeiten entfernt (probeweise) - ich hatte damit nicht immer ganz so viel Glück da die wallbox dann manchmal den alten / falschen Status meldet. Mit Wartezeiten schien das stabiler zu laufen, werde ich aber testen.

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,354 @@
+	return res, err
+}
+
+func (wb *Smaevcharger) getMeasurement(id string) (float64, error) {
```
Diese und die Folgefunktion könnte man auch noch an die Typen statt hier im Hauptteil hängen, das wäre aber optional.

powelllens committed:
fixed formatting
replaced switch with if/else -> only two options
simplify return
removed obsolete wb.reset()

powelllens committed:
wip

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,351 @@
+
+	req, err := request.New(http.MethodPut, uri, request.MarshalJSON(data), request.JSONEncoding)
+	if err == nil {
+		res, err := wb.Do(req)
```
Kannst du hier

    _, err = wb.DoBody(req)

nehmen? Das handelt auch gleich den Status mit und liest die Antwort (was notwendig ist).

powelllens committed:
changed Send from wb.Do(req) to wb.DoBody(req) for improved error handling

andig committed:
Update smaevcharger.go
